### PR TITLE
Add check_cacher.

### PR DESCRIPTION
**Pull request by chowchow316**

### CODE REVIEW

File: src/check_cacher.cc [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+  CacheElem* elem = lookup.value();
+
+  if (ShouldFlush(*elem)) {
+    // By returning NO_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  // Setting last check to now to block more check requests to Chemist.
+  elem->set_last_check_time(SimpleCycleTimer::Now());
```
Comment (qiwzhang) [inline src/check_cacher.cc]: set_last_check_time should be inside line 67

File: src/check_cacher.h [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+ private:
+  class CacheElem {
+   public:
+    CacheElem(const ::istio::mixer::v1::CheckResponse& response,
+              const int64_t time)
+        : check_response_(response), last_check_time_(time) {}
+
+    // Returns the Attributes and reset the cache entry.
+    Attributes ReturnAttributesAndClear();
```
Comment (qiwzhang) [inline src/check_cacher.h]: this is not needed.

File: src/check_cacher.h [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      const Attributes& request, ::istio::mixer::v1::CheckResponse* response);
+
+  // Caches a response from a remote mixer call.
+  virtual ::google::protobuf::util::Status CacheResponse(
+      const Attributes& request,
+      const ::istio::mixer::v1::CheckResponse& response);
+
+  // When the next Flush() should be called.
+  // Returns in ms from now, or -1 for never.
+  virtual int GetNextFlushInterval();
```
Comment (qiwzhang) [inline src/check_cacher.h]: this is not needed.

File: src/check_cacher.cc [hunk 4]
```diff
@@ -64,12 +64,12 @@ Status CheckCacher::Check(const Attributes& attributes,
   CacheElem* elem = lookup.value();
 
   if (ShouldFlush(*elem)) {
+    // Setting last check to now to block more check requests to Chemist.
```
Comment (qiwzhang) [inline src/check_cacher.cc]: s/Chemist/Mixer

File: src/check_cacher.h [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  std::unique_ptr<CheckCache> cache_;
+
+  // flush interval in cycles.
+  int64_t flush_interval_in_cycle_;
+
+  GOOGLE_DISALLOW_EVIL_CONSTRUCTORS(CheckCacher);
+};
+
+// Create a check cache.
+std::unique_ptr<CheckCacher> CreateCheckCacher(const CheckOptions& options);
```
Comment (qiwzhang) [inline src/check_cacher.h]: no need this Create function.  We will use CheckCache clas directly.

File: src/check_cacher.h [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+#include "mixer/api/v1/service.pb.h"
+#include "utils/simple_lru_cache.h"
+#include "utils/simple_lru_cache_inl.h"
+
+namespace istio {
+namespace mixer_client {
+
+// Cache Mixer Check Attributes.
+// This interface is thread safe.
+class CheckCacher {
```
Comment (qiwzhang) [inline src/check_cacher.h]: CheckCache

File: src/check_cacher.h [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+#ifndef MIXERCLIENT_CHECK_CACHE_H
+#define MIXERCLIENT_CHECK_CACHE_H
+
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "google/protobuf/stubs/status.h"
+#include "include/client.h"
```
Comment (qiwzhang) [inline src/check_cacher.h]: Should not need include/client.h

File: src/check_cacher.h (hunk 6 above)
Comment (chowchow316) [inline src/check_cacher.h]: Done. Renamed the other `CheckCache` to `CheckLRUCache.`

File: src/check_cacher.h (hunk 7 above)
Comment (chowchow316) [inline src/check_cacher.h]: `Attributes` is defined in `client.h`.

File: src/check_cacher.h (hunk 5 above)
Comment (chowchow316) [inline src/check_cacher.h]: Done.

**Pull request merged**